satellite/notification/service.go

```diff
@@ -0,0 +1,190 @@
+	//Clear map
+	service.limiter = map[string]ClientSetting{}
+	return nil
+}
+
+// Close closes the resources
+func (service *Service) Close() error {
+	service.loop.Stop()
+	service.loop.Close()
+
+	return nil
+}
+
+func (service *Service) IncrementLimiter(id string, email bool, rpc bool) {
+	service.lock.Lock()
+	defer service.lock.Unlock()
+
+	entry := service.limiter[id]
+	if email {
+		entry.Emails++
+	}
+	if rpc {
+		entry.RPC++
+	}
+	service.limiter[id] = entry
+}
+
+// CheckRPCLimit checks if hourly RPC limit've been reached.
+func (service *Service) CheckRPCLimit(id string) bool {
+	var ok bool
+	var entry ClientSetting
+
+	if entry, ok = service.limiter[id]; !ok {
+		return false
+	}
+
+	return entry.RPC < service.config.HourlyRPC
+}
+
+// CheckEmailLimit checks if hourly email limit've been reached.
+func (service *Service) CheckEmailLimit(id string) bool {
+	var ok bool
+	var entry ClientSetting
+
+	if entry, ok = service.limiter[id]; !ok {
+		return false
+	}
+
+	return entry.Emails < service.config.HourlyEmails
+}
+
+// ProcessNotification sends message to the specified set of nodes (ids).
+func (service *Service) ProcessNotification(ctx context.Context, message *pb.NotificationMessage) (err error) {
+	defer mon.Task()(&ctx)(&err)
+	var eSent, rSent = false, false
+
+	service.log.Debug("sending to node", zap.String("address", message.Address), zap.String("message", string(message.Message)))
+	if service.CheckRPCLimit(message.NodeId.String()) {
+		_, err = service.processNotificationRPC(ctx, message)
+		if err != nil {
+			return err
+		}
+		rSent = true
+	}
+	if service.CheckEmailLimit(message.NodeId.String()) {
+		err = service.processNotificationEmail(ctx, message)
+		if err != nil {
+			return err
+		}
+		eSent = true
+	}
+
+	service.IncrementLimiter(message.NodeId.String(), eSent, rSent)
+	return nil
+}
+
+// processNotificationRPC processing notification by rpc.
+func (service *Service) processNotificationRPC(ctx context.Context, message *pb.NotificationMessage) (_ *pb.NotificationResponse, err error) {
+	client, err := newClient(ctx, service.dialer, message.Address, message.NodeId)
+	if err != nil {
+		// if this is a network error, then return the error otherwise just report internal error
+		_, ok := err.(net.Error)
+		if ok {
+			return &pb.NotificationResponse{}, Error.New("failed to connect to %s: %v", message.Address, err)
+		}
+
+		service.log.Warn("internal error", zap.String("error", err.Error()))
+		return &pb.NotificationResponse{}, Error.New("couldn't connect to client at addr: %s due to internal error.", message.Address)
+	}
+	defer func() { err = errs.Combine(err, client.Close()) }()
+
+	return client.client.ProcessNotification(ctx, message)
+}
+
+// processNotificationEmail processing notification by mail service.
+func (service *Service) processNotificationEmail(ctx context.Context, message *pb.NotificationMessage) (err error) {
+	//return endpoint.service.mailer.Send(ctx, &post.Message{})
```

Comment:
This needs to be plugged in with the actual email template, but is fine to merge without it for now